In [1]:
%load_ext nb_black

<IPython.core.display.Javascript object>

In [2]:
import requests
import pandas as pd
import numpy as np
import sqlalchemy as sqla
import plotnine as p9
import seaborn as sns

<IPython.core.display.Javascript object>

In [3]:
# Request hesitancy data
hesitancy_endpoint = "https://data.cdc.gov/resource/q9mh-h2tw.json?$limit=4000"
r = requests.get(hesitancy_endpoint)
hesitancy_df = pd.DataFrame(r.json())
hesitancy_df.head()

fips_code                 county_name    state estimated_hesitant  \
0      1123  Tallapoosa County, Alabama  ALABAMA             0.1806   
1      1121   Talladega County, Alabama  ALABAMA             0.1783   
2      1131      Wilcox County, Alabama  ALABAMA             0.1735   
3      1129  Washington County, Alabama  ALABAMA             0.1735   
4      1133     Winston County, Alabama  ALABAMA             0.1805   

  estimated_hesitant_or_unsure estimated_strongly_hesitant  \
0                         0.24                      0.1383   
1                        0.235                      0.1368   
2                       0.2357                      0.1337   
3                       0.2357                      0.1337   
4                       0.2313                      0.1379   

  social_vulnerability_index             svi_category  \
0                       0.89  Very High Vulnerability   
1                       0.87  Very High Vulnerability   
2                       0.93  Very High Vulnerability   
3                       0.73       High Vulnerability   
4                        0.7       High Vulnerability   

  ability_to_handle_a_covid      cvac_category  ...  \
0                      0.64       High Concern  ...   
1                      0.84  Very High Concern  ...   
2                      0.94  Very High Concern  ...   
3                      0.82  Very High Concern  ...   
4                       0.8       High Concern  ...   

  percent_non_hispanic_asian percent_non_hispanic_black  \
0                     0.0036                     0.2697   
1                     0.0061                     0.3237   
2                     0.0003                     0.6938   
3                     0.0025                     0.2354   
4                     0.0016                     0.0073   

  percent_non_hispanic_native percent_non_hispanic_white  \
0                           0                     0.6887   
1                      0.0003                     0.6263   
2                           0                     0.2684   
3                           0                     0.6495   
4                      0.0005                      0.937   

                                  geographical_point state_code  \
0  {'type': 'Point', 'coordinates': [-86.844516, ...         AL   
1  {'type': 'Point', 'coordinates': [-86.844516, ...         AL   
2  {'type': 'Point', 'coordinates': [-86.844516, ...         AL   
3  {'type': 'Point', 'coordinates': [-86.844516, ...         AL   
4  {'type': 'Point', 'coordinates': [-86.844516, ...         AL   

                                     county_boundary  \
0  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
1  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
2  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
3  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
4  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   

                                      state_boundary  \
0  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
1  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
2  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
3  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
4  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   

  :@computed_region_hjsp_umg2 :@computed_region_skr5_azej  
0                          29                          94  
1                          29                          94  
2                          29                          94  
3                          29                          94  
4                          29                          94  

[5 rows x 23 columns]

<IPython.core.display.Javascript object>

In [4]:
# Extracting political party data from our public github
url = "https://raw.githubusercontent.com/gspiga/STAT141Bfinal/main/data/raw_data.csv"
poli_df = pd.read_csv(url)
poli_df.head()

Location Governor Political Affiliation  \
0     Alabama                     Republican   
1      Alaska                     Republican   
2     Arizona                     Republican   
3    Arkansas                     Republican   
4  California                       Democrat   

  State Senate Majority Political Affiliation  \
0                                  Republican   
1                                  Republican   
2                                  Republican   
3                                  Republican   
4                                    Democrat   

  State House Majority Political Affiliation  \
0                                 Republican   
1                                 Republican   
2                                 Republican   
3                                 Republican   
4                                   Democrat   

  State Attorney General Political Affiliation  \
0                                   Republican   
1                                    Appointed   
2                                   Republican   
3                                   Republican   
4                                     Democrat   

  State Insurance Commissioner Political Affiliation  Footnotes  
0                                          Appointed        NaN  
1                                          Appointed        NaN  
2                                          Appointed        NaN  
3                                          Appointed        NaN  
4                                           Democrat        NaN

<IPython.core.display.Javascript object>

In [5]:
# Changing one column in poli_df to upper case so that we may join them (pandas is case sensitive).
poli_df["Location"] = poli_df["Location"].apply(
    lambda x: x.upper()  # Uppercase all state names
)
poli_df.head()

Location Governor Political Affiliation  \
0     ALABAMA                     Republican   
1      ALASKA                     Republican   
2     ARIZONA                     Republican   
3    ARKANSAS                     Republican   
4  CALIFORNIA                       Democrat   

  State Senate Majority Political Affiliation  \
0                                  Republican   
1                                  Republican   
2                                  Republican   
3                                  Republican   
4                                    Democrat   

  State House Majority Political Affiliation  \
0                                 Republican   
1                                 Republican   
2                                 Republican   
3                                 Republican   
4                                   Democrat   

  State Attorney General Political Affiliation  \
0                                   Republican   
1                                    Appointed   
2                                   Republican   
3                                   Republican   
4                                     Democrat   

  State Insurance Commissioner Political Affiliation  Footnotes  
0                                          Appointed        NaN  
1                                          Appointed        NaN  
2                                          Appointed        NaN  
3                                          Appointed        NaN  
4                                           Democrat        NaN

<IPython.core.display.Javascript object>

In [21]:
# We join the dataframe
polihes_df = hesitancy_df.merge(poli_df, left_on="state", right_on="Location")

# Have to change decimal values from string to numeric
cols = [
    "estimated_hesitant",
    "estimated_hesitant_or_unsure",
    "estimated_strongly_hesitant",
    "estimated_strongly_hesitant",
    "social_vulnerability_index",
    "ability_to_handle_a_covid",
    "percent_hispanic",
    "percent_non_hispanic_american",
    "percent_non_hispanic_asian",
    "percent_non_hispanic_black",
    "percent_non_hispanic_native",
    "percent_non_hispanic_white",
]
polihes_df[cols] = polihes_df[cols].apply(pd.to_numeric, errors="coerce", axis=1)

<IPython.core.display.Javascript object>

In [7]:
polihes_df.head()

fips_code                 county_name    state  estimated_hesitant  \
0      1123  Tallapoosa County, Alabama  ALABAMA              0.1806   
1      1121   Talladega County, Alabama  ALABAMA              0.1783   
2      1131      Wilcox County, Alabama  ALABAMA              0.1735   
3      1129  Washington County, Alabama  ALABAMA              0.1735   
4      1133     Winston County, Alabama  ALABAMA              0.1805   

   estimated_hesitant_or_unsure  estimated_strongly_hesitant  \
0                        0.2400                       0.1383   
1                        0.2350                       0.1368   
2                        0.2357                       0.1337   
3                        0.2357                       0.1337   
4                        0.2313                       0.1379   

   social_vulnerability_index             svi_category  \
0                        0.89  Very High Vulnerability   
1                        0.87  Very High Vulnerability   
2                        0.93  Very High Vulnerability   
3                        0.73       High Vulnerability   
4                        0.70       High Vulnerability   

   ability_to_handle_a_covid      cvac_category  ...  \
0                       0.64       High Concern  ...   
1                       0.84  Very High Concern  ...   
2                       0.94  Very High Concern  ...   
3                       0.82  Very High Concern  ...   
4                       0.80       High Concern  ...   

                                      state_boundary  \
0  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
1  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
2  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
3  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   
4  {'type': 'MultiPolygon', 'coordinates': [[[[-8...   

   :@computed_region_hjsp_umg2 :@computed_region_skr5_azej  Location  \
0                           29                          94   ALABAMA   
1                           29                          94   ALABAMA   
2                           29                          94   ALABAMA   
3                           29                          94   ALABAMA   
4                           29                          94   ALABAMA   

   Governor Political Affiliation  \
0                      Republican   
1                      Republican   
2                      Republican   
3                      Republican   
4                      Republican   

   State Senate Majority Political Affiliation  \
0                                   Republican   
1                                   Republican   
2                                   Republican   
3                                   Republican   
4                                   Republican   

   State House Majority Political Affiliation  \
0                                  Republican   
1                                  Republican   
2                                  Republican   
3                                  Republican   
4                                  Republican   

  State Attorney General Political Affiliation  \
0                                   Republican   
1                                   Republican   
2                                   Republican   
3                                   Republican   
4                                   Republican   

  State Insurance Commissioner Political Affiliation Footnotes  
0                                          Appointed       NaN  
1                                          Appointed       NaN  
2                                          Appointed       NaN  
3                                          Appointed       NaN  
4                                          Appointed       NaN  

[5 rows x 30 columns]

<IPython.core.display.Javascript object>

We will now form a linear regression model based on SVI, Death rate, ability to handle a covid outbreak, and governor political affiliation to predict hesistancy. 

In [8]:
import statsmodels.api as sm

<IPython.core.display.Javascript object>

In [27]:
X_eth = polihes_df[
    [
        "percent_non_hispanic_american",
        "percent_non_hispanic_asian",
        "percent_hispanic",
        "percent_non_hispanic_black",
        "percent_non_hispanic_native",
        "percent_non_hispanic_white",
    ]
]
X_eth = sm.add_constant(X_eth)
Y = polihes_df["estimated_hesitant"]

model = sm.OLS(Y, X_eth).fit()
predictions = model.predict(X_eth)

model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:     estimated_hesitant   R-squared:                       0.165
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     103.5
Date:                Wed, 08 Dec 2021   Prob (F-statistic):          2.86e-119
Time:                        22:55:24   Log-Likelihood:                 5477.9
No. Observations:                3142   AIC:                        -1.094e+04
Df Residuals:                    3135   BIC:                        -1.090e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
=================================================================================================
                                    coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------
const                             0.3897      0.051      7.702      0.000       0.290       0.489
percent_non_hispanic_american    -0.1629      0.055     -2.962      0.003      -0.271      -0.055
percent_non_hispanic_asian       -0.7552      0.067    -11.246      0.000      -0.887      -0.624
percent_hispanic                 -0.3045      0.051     -5.944      0.000      -0.405      -0.204
percent_non_hispanic_black       -0.2031      0.052     -3.942      0.000      -0.304      -0.102
percent_non_hispanic_native      -0.0342      0.163     -0.210      0.833      -0.353       0.285
percent_non_hispanic_white       -0.2583      0.052     -5.011      0.000      -0.359      -0.157
==============================================================================
Omnibus:                      113.032   Durbin-Watson:                   0.296
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              152.412
Skew:                           0.377   Prob(JB):                     8.02e-34
Kurtosis:                       3.772   Cond. No.                         312.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

<IPython.core.display.Javascript object>

In [30]:
# Y_sh = polihes_df["estimated_strongly_hesitant"]

# model2 = sm.OLS(Y_sh, X_eth).fit()
# predictions = model2.predict(X_eth)

# model2.summary()

<IPython.core.display.Javascript object>

From our model summary we can see that we get a model for predicting hesistancy with equation $y = 0.3897 - 0.1629x_{nhamer}- -0.7552x_{asian} - 0.3045x_{hispan} - 0.2031x_{black} - 0.0342x_{natam} - 0.2583x_{white}$ We can see that the ethnicity population with the largest influence on vaccine hesitancy is non-hispanic asians with a coefficient of -0.7552. As with all the coefficients in this model, the negative values represent a relationship corresponding to the increase in population of a certain ethnicity/race lowers the estimated hesitancy of a population. 

Another factor to notice in the model is our adjusted R-squared value, which is only 0.164. 